roomserver/storage/shared/storage.go

```diff
@@ -1589,6 +1591,29 @@ func (d *Database) UpgradeRoom(ctx context.Context, oldRoomID, newRoomID, eventS
 	})
 }
 
+// InsertUserRoomKey inserts a new user room key for the given user and room.
+// Returns an error if a database error occurred, also if the primary constraint was violated.
+func (d *Database) InsertUserRoomKey(ctx context.Context, userNID types.EventStateKeyNID, roomNID types.RoomNID, key ed25519.PrivateKey) error {
+	return d.Writer.Do(d.DB, nil, func(txn *sql.Tx) error {
+		return d.UserRoomKeyTable.InsertUserRoomKey(ctx, txn, userNID, roomNID, key)
+	})
+}
+
+// SelectUserRoomKey queries the user room key for a given user.
+// Returns the key and an error.
+// TODO: should we handle absent keys (sql.ErrNoRows) as non-fatal?
+func (d *Database) SelectUserRoomKey(ctx context.Context, userNID types.EventStateKeyNID, roomNID types.RoomNID) (key ed25519.PrivateKey, err error) {
```

Comment:
It's not entirely clear that this is for selecting the private key i.e because Alice wants to send a message and now you need to sign an event, as opposed to "here's a set of events you're about to return to the client, please pull out the user IDs for these sender keys". It's not clear because of comments like:

> queries the user room key for a given user

It's not clear if this is the public part or private part.

> TODO: should we handle absent keys (sql.ErrNoRows) as non-fatal?

Yes, and let the caller decide what to do if the user has no key (maybe it can make one).